src/solve/fulfill.rs

```diff
@@ -217,67 +241,119 @@ impl<'s> Fulfill<'s> {
                 }
             }
 
-            self.obligations.extend(obligations.drain(..));
-            debug!("end of round, {} obligations left", self.obligations.len());
+            self.to_prove.extend(obligations.drain(..));
+            debug!("end of round, {} obligations left", self.to_prove.len());
         }
 
-        // At the end of this process, `self.obligations` should have
+        // At the end of this process, `self.to_prove` should have
         // all of the ambiguous obligations, and `obligations` should
         // be empty.
         assert!(obligations.is_empty());
 
-        if self.obligations.is_empty() {
-            // If no obligations remain, we have definitively solved our goals,
-            // and can return the results of inference as a unique substitution
-            let constraints = self.constraints.into_iter().collect();
-            let constrained = self.infer.canonicalize(&ConstrainedSubst { subst, constraints });
-            Ok(Solution::Unique(constrained.quantified))
+        if self.to_prove.is_empty() {
+            Ok(Outcome::Complete)
         } else {
-            // If we have obligations remaining, we need to determine how to
-            // package up what we learned about type inference as an ambiguous
-            // solution.
-
-            if subst.is_trivial(&mut self.infer) {
-                // In this case, we didn't learn *anything* definitively. So
-                // now, we go one last time through the obligations, this time
-                // applying even *tentative* inference suggestions, so that we
-                // can yield these upwards as our own suggestions. In
-                // particular, we yield up the first one we can find.
-
-                while let Some(wc) = self.obligations.pop() {
-                    let (free_vars, solution) = self.solve_obligation(&wc).unwrap();
-                    if let Some(constrained_subst) = solution.constrained_subst() {
-                        self.apply_solution(free_vars, constrained_subst);
-                        let subst = self.infer.canonicalize(&subst);
-                        return Ok(Solution::Ambig(Guidance::Suggested(subst.quantified)));
-                    }
+            Ok(Outcome::Incomplete)
+        }
+    }
+
+    fn fulfill_negative(&mut self) -> Result<Outcome> {
+        // Attempt to refute each negative obligations by in a fresh Fulfill
+        for goal in self.to_refute.drain(..) {
+            let canonicalized = self.infer.canonicalize(&goal);
+            let goal = canonicalized.quantified.value;
+
+            // Negation cannot be used to resolve existential variables, and do
+            // not have a useful (for us) logial meaning when they contain
+            // existential variables; treat negative goals with free variables
+            // as ambiguous
+            if !canonicalized.free_vars.is_empty() {
+                return Ok(Outcome::Incomplete);
+            }
```

Comment:
This is stronger than necessary, right? For example, imagine `Vec<?T>: Copy` -- given that there are no impls of `Copy` for `Vec` at all, and assuming a closed world, we could figure out that this is false. In particular, if we get a `Unique` result with an empty substitution, we are satisfied, even if `free_vars` is non-empty...?